### PR TITLE
8311939: Excessive allocation of Matcher.groups array

### PR DESCRIPTION
**Pull request by deathy**
Reduces excessive allocation of Matcher.groups array when the original Pattern has no groups or less than 9 groups.

Original clamping to 10 possibly due to documented behavior from javadoc: 
"In this class, \1 through \9 are always interpreted as back references, "

Only with Matcher changes RegExTest.backRefTest fails when backreferences to non-existing groups are present.
Added a match failure condition in Pattern that fixes failing tests.

As per existing `java.util.regex.Pattern.BackRef#match`: "// If the referenced group didn't match, neither can this"

A group that does not exist in the original Pattern can never match so neither can a backref to that group.
If the group existed in the original Pattern then it would have had space allocated in Matcher.groups for that group index.
So a group index outside groups array length must never match.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8311939](https://bugs.openjdk.org/browse/JDK-8311939): Excessive allocation of Matcher.groups array (**Enhancement** - P4)

### Reviewers
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)
 * [Ian Graves](https://openjdk.org/census#igraves) (@igraves - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14894/head:pull/14894` \
`$ git checkout pull/14894`

Update a local copy of the PR: \
`$ git checkout pull/14894` \
`$ git pull https://git.openjdk.org/jdk.git pull/14894/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14894`

View PR using the GUI difftool: \
`$ git pr show -t 14894`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14894.diff">https://git.openjdk.org/jdk/pull/14894.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14894#issuecomment-1640207310)

### CODE REVIEW

**Comment by deathy** [comment]
updated to also reduce allocation in `java.util.regex.Matcher#usePattern`

all jdk_util tests passing

ran JMH `java.util.regex.FindPattern` test and times seem better but test is pretty light (large errors compared to avg score)

ran JMH with `-prof gc`:

```
Benchmark                                 (patternString)                                      (text)  Mode  Cnt    Score    Error   Units
FindPattern.testFind:·gc.alloc.rate.norm     [^A-Za-z0-9]  abcdefghijklmnop1234567890ABCDEFGHIJKLMNOP  avgt    3  207.999 ±  0.030    B/op
FindPattern.testFind:·gc.alloc.rate.norm     [^A-Za-z0-9]  ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,  avgt    3  208.000 ±  0.001    B/op
FindPattern.testFind:·gc.alloc.rate.norm      [A-Za-z0-9]  abcdefghijklmnop1234567890ABCDEFGHIJKLMNOP  avgt    3  207.861 ±  4.395    B/op
FindPattern.testFind:·gc.alloc.rate.norm      [A-Za-z0-9]  ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,  avgt    3  207.999 ±  0.031    B/op


Benchmark                                 (patternString)                                      (text)  Mode  Cnt    Score     Error   Units
FindPattern.testFind:·gc.alloc.rate.norm     [^A-Za-z0-9]  abcdefghijklmnop1234567890ABCDEFGHIJKLMNOP  avgt    3   56.181 ±   5.713    B/op
FindPattern.testFind:·gc.alloc.rate.norm     [^A-Za-z0-9]  ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,  avgt    3  136.000 ±   0.001    B/op
FindPattern.testFind:·gc.alloc.rate.norm      [A-Za-z0-9]  abcdefghijklmnop1234567890ABCDEFGHIJKLMNOP  avgt    3   56.000 ±   0.010    B/op
FindPattern.testFind:·gc.alloc.rate.norm      [A-Za-z0-9]  ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,  avgt    3  135.999 ±   0.028    B/op
```


As expected 72 byte/op reduction for case when pattern doesn't match.
Unexpected: seems double reduction in allocations for the case when pattern matches. Not completely sure where that is coming from. Maybe some optimizations in loops since these don't have groups and array is always 2 elements?

**Comment by deathy** [comment]
ping @shipilev, if you could take a look here

File: src/java.base/share/classes/java/util/regex/Pattern.java [hunk 1]
```diff
@@ -5187,6 +5187,13 @@ static class BackRef extends Node {
             groupIndex = groupCount + groupCount;
         }
         boolean match(Matcher matcher, int i, CharSequence seq) {
+
```
Comment (shipilev) [inline src/java.base/share/classes/java/util/regex/Pattern.java:5190]: Excess new line.

**Review by shipilev — COMMENTED** [review]
Shouldn't the similar change be in `CIBackRef.match` too? The fact current tests do not catch it makes me uneasy: the test coverage seems to be rather low there. 

We need a regex expert to look at it. @rgiulietti @igraves might help us out here?

**Comment by deathy** [comment]
`RegExTest#backRefTest` seems to be pretty extensive but only for `BackRef` not `CIBackRef`

I saw one `CIBackRef` related test added in https://github.com/openjdk/jdk/pull/7501 but it's very simple/specific.

I triggered test failure locally by duplicating `backRefTest` 1-9 loop with `(?i)` in pattern so yes it seems like `CIBackRef` needs same change.

But not sure about the test, duplicating loop seems odd. Maybe entire `RegExTest#backRefTest` needs to be duplicated into a `ciBackRefTest` with all patterns preprended with `(?i)` ? Wouldn't look too clean.

File: src/java.base/share/classes/java/util/regex/Pattern.java (hunk 1 above)
Comment (deathy) [reply src/java.base/share/classes/java/util/regex/Pattern.java]: line removed

**Comment by deathy** [comment]
Made changes also in `CIBackRef` and copied/changed test into new `ciBackRefTest`

Not pretty since changes to one could miss the other, but all patterns are different. (for what it's worth tests pass...)

There's also a special case with supplementary character tests since `toSupplementaries` given `(?i)` generates invalid pattern, so I had to change those like:
``
pattern = Pattern.compile("(?i)" + toSupplementaries("(a*)bc\\1"));
``

Definitely needs a close look by a regex expert.

File: src/java.base/share/classes/java/util/regex/Pattern.java [hunk 2]
```diff
@@ -5187,6 +5187,12 @@ static class BackRef extends Node {
             groupIndex = groupCount + groupCount;
         }
         boolean match(Matcher matcher, int i, CharSequence seq) {
+            // reference to not existing group must never match
+            // group does not exist if matcher didn't allocate space for it
+            if (groupIndex + 1 > matcher.groups.length) {
```
Comment (rgiulietti) [inline src/java.base/share/classes/java/util/regex/Pattern.java:5192]:
```suggestion
            if (groupIndex >= matcher.groups.length) {
```

File: src/java.base/share/classes/java/util/regex/Pattern.java [hunk 3]
```diff
@@ -5223,6 +5229,12 @@ static class CIBackRef extends Node {
             this.doUnicodeCase = doUnicodeCase;
         }
         boolean match(Matcher matcher, int i, CharSequence seq) {
+            // reference to not existing group must never match
+            // group does not exist if matcher didn't allocate space for it
+            if (groupIndex + 1 > matcher.groups.length) {
```
Comment (rgiulietti) [inline src/java.base/share/classes/java/util/regex/Pattern.java:5234]:
```suggestion
            if (groupIndex >= matcher.groups.length) {
```

**Review by rgiulietti — COMMENTED** [review]
The change looks good.
However, I'm not a Reviewer.

**Comment by deathy** [comment]
Thanks. I committed changes suggested for if conditions.

**Comment by deathy** [comment]
@shipilev anything I should do to continue here?

**Comment by shipilev** [comment]
> @shipilev anything I should do to continue here?

Not really, we need more reviewers.

**Comment by igraves** [comment]
This looks good to me, too. Like @rgiulietti I'm also not a reviewer.

Tagging in @stuart-marks and @RogerRiggs for reviewer status.

**Comment by shipilev** [comment]
@deathy, please merge from master to get the clean testing.

**Comment by RogerRiggs** [comment]
Looks good.
Non-"R"-reviewers can review and approve (and get recognition for reviewing).

**Review by RogerRiggs — APPROVED** [review]
LGTM

**Pull request closed without merging**